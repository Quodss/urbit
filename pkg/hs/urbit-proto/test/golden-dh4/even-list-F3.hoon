=>
|%
++  list
  ^-  $
  $%  [%nil $~]
      [%con @ list]
  ==
++  even-list
  ^-  $
  $%  [%nil $~]
      [%con @ $con @ even-list]  :: FIXME allow %con
  ==
--
`even-list``list`[%nil ~]
