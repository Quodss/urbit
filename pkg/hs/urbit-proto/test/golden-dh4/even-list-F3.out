PARSED

=>  |%
    ++  even-list
      ^-  $
      $%  {$nil [$~]}
          {$con [@ $con @ even-list]}
      ==
    ++  list
      ^-  $
      $%  {$nil [$~]}
          {$con [@ list]}
      ==
    --
`even-list``list`[%nil ~]

SOFT

=>  |%
    ++  even-list
      ^-  $
      $:  ?(%7104878 %7237475)
          ^-  $
          ?:  ?=(%"nil" +2)
            $~
          ??  ?=(%"con" +2)
          [@ $con @ even-list]
      ==
    ++  list
      ^-  $
      $:  ?(%7104878 %7237475)
          ^-  $
          ?:  ?=(%"nil" +2)
            $~
          ??  ?=(%"con" +2)
          [@ list]
      ==
    --
`even-list``list`[%nil ~]

ERROR

  trace:  play:
          lvl   0
          sut   0|*
          ----
          code  =>  |%
                    ++  even-list
                      ^-  $
                      $:  ?(%7104878 %7237475)
                          ^-  $
                          ?:  ?=(%"nil" +2)
                            $~
                          ??  ?=(%"con" +2)
                          [@ $con @ even-list]
                      ==
                    ++  list
                      ^-  $
                      $:  ?(%7104878 %7237475)
                          ^-  $
                          ?:  ?=(%"nil" +2)
                            $~
                          ??  ?=(%"con" +2)
                          [@ list]
                      ==
                    --
                `even-list``list`[%nil ~]


          play:
          lvl   0
          sut   $=  =>  0
                    |%
                    ++  even-list
                      $:  ?(%7104878 %7237475)
                          ?:  ?=(%"nil" +2)
                            $~
                          [@ $con @ even-list.+3]
                      ==
                    ++  list
                      $:  ?(%7104878 %7237475)
                          ?:  ?=(%"nil" +2)
                            $~
                          [@ list.+3]
                      ==
                    --
                $|  0|*
                  0|*
                ++  even-list  $
                ++  list  $
                --

          ----
          code  `even-list``list`[%nil ~]

          work:
          lvl   0
          sut   $=  =>  0
                    |%
                    ++  even-list
                      $:  ?(%7104878 %7237475)
                          ?:  ?=(%"nil" +2)
                            $~
                          [@ $con @ even-list.+3]
                      ==
                    ++  list
                      $:  ?(%7104878 %7237475)
                          ?:  ?=(%"nil" +2)
                            $~
                          [@ list.+3]
                      ==
                    --
                $|  0|*
                  0|*
                ++  even-list  $
                ++  list  $
                --

          ----
          mode  cast
          code  `list`[%nil ~]
          type  =>  =>  0
                    |%
                    ++  even-list
                      $:  ?(%7104878 %7237475)
                          ?:  ?=(%"nil" +2)
                            $~
                          [@ $con @ even-list.+3]
                      ==
                    ++  list
                      $:  ?(%7104878 %7237475)
                          ?:  ?=(%"nil" +2)
                            $~
                          [@ list.+3]
                      ==
                    --
                $:  ?(%7104878 %7237475)
                    ?:  ?=(%"nil" +2)
                      $~
                    [@ $con @ even-list.+3]
                ==


          cast:
          have  =>  =>  0
                    |%
                    ++  even-list
                      $:  ?(%7104878 %7237475)
                          ?:  ?=(%"nil" +2)
                            $~
                          [@ $con @ even-list.+3]
                      ==
                    ++  list
                      $:  ?(%7104878 %7237475)
                          ?:  ?=(%"nil" +2)
                            $~
                          [@ list.+3]
                      ==
                    --
                $:  ?(%7104878 %7237475)
                    ?:  ?=(%"nil" +2)
                      $~
                    [@ list.+3]
                ==

          need  =>  =>  0
                    |%
                    ++  even-list
                      $:  ?(%7104878 %7237475)
                          ?:  ?=(%"nil" +2)
                            $~
                          [@ $con @ even-list.+3]
                      ==
                    ++  list
                      $:  ?(%7104878 %7237475)
                          ?:  ?=(%"nil" +2)
                            $~
                          [@ list.+3]
                      ==
                    --
                $:  ?(%7104878 %7237475)
                    ?:  ?=(%"nil" +2)
                      $~
                    [@ $con @ even-list.+3]
                ==


          cast:
          have  $:  ?(%7104878 %7237475)
                    ?:  ?=(%"nil" +2)
                      $~
                    :-  @
                    =>  =>  0
                        |%
                        ++  even-list
                          $:  ?(%7104878 %7237475)
                              ?:  ?=(%"nil" +2)
                                $~
                              [@ $con @ even-list.+3]
                          ==
                        ++  list
                          $:  ?(%7104878 %7237475)
                              ?:  ?=(%"nil" +2)
                                $~
                              [@ list.+3]
                          ==
                        --
                    $:  ?(%7104878 %7237475)
                        ?:  ?=(%"nil" +2)
                          $~
                        [@ list.+3]
                    ==
                ==

          need  =>  =>  0
                    |%
                    ++  even-list
                      $:  ?(%7104878 %7237475)
                          ?:  ?=(%"nil" +2)
                            $~
                          [@ $con @ even-list.+3]
                      ==
                    ++  list
                      $:  ?(%7104878 %7237475)
                          ?:  ?=(%"nil" +2)
                            $~
                          [@ list.+3]
                      ==
                    --
                $:  ?(%7104878 %7237475)
                    ?:  ?=(%"nil" +2)
                      $~
                    [@ $con @ even-list.+3]
                ==


          cast:
          have  $:  ?(%7104878 %7237475)
                    ?:  ?=(%"nil" +2)
                      $~
                    :-  @
                    =>  =>  0
                        |%
                        ++  even-list
                          $:  ?(%7104878 %7237475)
                              ?:  ?=(%"nil" +2)
                                $~
                              [@ $con @ even-list.+3]
                          ==
                        ++  list
                          $:  ?(%7104878 %7237475)
                              ?:  ?=(%"nil" +2)
                                $~
                              [@ list.+3]
                          ==
                        --
                    $:  ?(%7104878 %7237475)
                        ?:  ?=(%"nil" +2)
                          $~
                        [@ list.+3]
                    ==
                ==

          need  $:  ?(%7104878 %7237475)
                    ?:  ?=(%"nil" +2)
                      $~
                    :^  @
                        $con
                      @
                    =>  =>  0
                        |%
                        ++  even-list
                          $:  ?(%7104878 %7237475)
                              ?:  ?=(%"nil" +2)
                                $~
                              [@ $con @ even-list.+3]
                          ==
                        ++  list
                          $:  ?(%7104878 %7237475)
                              ?:  ?=(%"nil" +2)
                                $~
                              [@ list.+3]
                          ==
                        --
                    $:  ?(%7104878 %7237475)
                        ?:  ?=(%"nil" +2)
                          $~
                        [@ $con @ even-list.+3]
                    ==
                ==


          cast:
          have  ?:  ?=(%"nil" +2)
                  $~
                :-  @
                =>  =>  0
                    |%
                    ++  even-list
                      $:  ?(%7104878 %7237475)
                          ?:  ?=(%"nil" +2)
                            $~
                          [@ $con @ even-list.+3]
                      ==
                    ++  list
                      $:  ?(%7104878 %7237475)
                          ?:  ?=(%"nil" +2)
                            $~
                          [@ list.+3]
                      ==
                    --
                $:  ?(%7104878 %7237475)
                    ?:  ?=(%"nil" +2)
                      $~
                    [@ list.+3]
                ==

          need  ?:  ?=(%"nil" +2)
                  $~
                :^  @
                    $con
                  @
                =>  =>  0
                    |%
                    ++  even-list
                      $:  ?(%7104878 %7237475)
                          ?:  ?=(%"nil" +2)
                            $~
                          [@ $con @ even-list.+3]
                      ==
                    ++  list
                      $:  ?(%7104878 %7237475)
                          ?:  ?=(%"nil" +2)
                            $~
                          [@ list.+3]
                      ==
                    --
                $:  ?(%7104878 %7237475)
                    ?:  ?=(%"nil" +2)
                      $~
                    [@ $con @ even-list.+3]
                ==


          cast:
          have  :-  @
                =>  =>  0
                    |%
                    ++  even-list
                      $:  ?(%7104878 %7237475)
                          ?:  ?=(%"nil" +2)
                            $~
                          [@ $con @ even-list.+3]
                      ==
                    ++  list
                      $:  ?(%7104878 %7237475)
                          ?:  ?=(%"nil" +2)
                            $~
                          [@ list.+3]
                      ==
                    --
                $:  ?(%7104878 %7237475)
                    ?:  ?=(%"nil" +2)
                      $~
                    [@ list.+3]
                ==

          need  :^  @
                    $con
                  @
                =>  =>  0
                    |%
                    ++  even-list
                      $:  ?(%7104878 %7237475)
                          ?:  ?=(%"nil" +2)
                            $~
                          [@ $con @ even-list.+3]
                      ==
                    ++  list
                      $:  ?(%7104878 %7237475)
                          ?:  ?=(%"nil" +2)
                            $~
                          [@ list.+3]
                      ==
                    --
                $:  ?(%7104878 %7237475)
                    ?:  ?=(%"nil" +2)
                      $~
                    [@ $con @ even-list.+3]
                ==


          cast:
          have  =>  =>  0
                    |%
                    ++  even-list
                      $:  ?(%7104878 %7237475)
                          ?:  ?=(%"nil" +2)
                            $~
                          [@ $con @ even-list.+3]
                      ==
                    ++  list
                      $:  ?(%7104878 %7237475)
                          ?:  ?=(%"nil" +2)
                            $~
                          [@ list.+3]
                      ==
                    --
                $:  ?(%7104878 %7237475)
                    ?:  ?=(%"nil" +2)
                      $~
                    [@ list.+3]
                ==

          need  :+  $con
                  @
                =>  =>  0
                    |%
                    ++  even-list
                      $:  ?(%7104878 %7237475)
                          ?:  ?=(%"nil" +2)
                            $~
                          [@ $con @ even-list.+3]
                      ==
                    ++  list
                      $:  ?(%7104878 %7237475)
                          ?:  ?=(%"nil" +2)
                            $~
                          [@ list.+3]
                      ==
                    --
                $:  ?(%7104878 %7237475)
                    ?:  ?=(%"nil" +2)
                      $~
                    [@ $con @ even-list.+3]
                ==


          cast:
          have  $:  ?(%7104878 %7237475)
                    ?:  ?=(%"nil" +2)
                      $~
                    :-  @
                    =>  =>  0
                        |%
                        ++  even-list
                          $:  ?(%7104878 %7237475)
                              ?:  ?=(%"nil" +2)
                                $~
                              [@ $con @ even-list.+3]
                          ==
                        ++  list
                          $:  ?(%7104878 %7237475)
                              ?:  ?=(%"nil" +2)
                                $~
                              [@ list.+3]
                          ==
                        --
                    $:  ?(%7104878 %7237475)
                        ?:  ?=(%"nil" +2)
                          $~
                        [@ list.+3]
                    ==
                ==

          need  :+  $con
                  @
                =>  =>  0
                    |%
                    ++  even-list
                      $:  ?(%7104878 %7237475)
                          ?:  ?=(%"nil" +2)
                            $~
                          [@ $con @ even-list.+3]
                      ==
                    ++  list
                      $:  ?(%7104878 %7237475)
                          ?:  ?=(%"nil" +2)
                            $~
                          [@ list.+3]
                      ==
                    --
                $:  ?(%7104878 %7237475)
                    ?:  ?=(%"nil" +2)
                      $~
                    [@ $con @ even-list.+3]
                ==


          cast:
          have  ?(%7104878 %7237475)
          need  $con


  cast-fail:
  have  ?(%7104878 %7237475)
  need  $con

